Review adding AD as an external authentication source

asteflova opened the pull request:
### What changes are you introducing?

After reviewing the existing procedure for adding Active Directory as an external authentication source when the base system of the Foreman server is connected directly to AD, I propose these changes:

* Drop steps related to privilege separation for Apache
* Drop a couple of other instructions that are now obsolete
* Drop adding more detailed explanations of the steps the procedure recommends
* Add a list of authentication features available to AD users
* Other edits for clarity and to set the context of the whole scenario

~~The current procedure for direct AD integration + configuring AD as an external authentication source + using GSS Proxy for privilege separation is faulty. For example, the steps don't really achieve Apache keytab separation as promised, they combine Samba-based and SSSD-based AD join, and some of them could be replaced with a reference to FreeIPA/IdM docs to make sure we don't need to fret about keeping in sync with SSSD development.~~

~~In its latest state (after quite a few changes in course), this PR proposes a complete restart to how we document the use case of direct AD integration -- and to document a minimum viable workflow to achieve direct AD integration and adding AD as an external authentication source:~~

~~* First, refer users to FreeIPA/IdM docs for the integration to AD (referencing SSSD as [the recommended way to join an AD domain](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html-single/integrating_rhel_systems_directly_with_windows_active_directory/index#overview-of-direct-integration-using-sssd_connecting-rhel-systems-directly-to-ad-using-sssd) and omitting references to Samba)~~
~~* Then, advise users on how to configure AD as an external authentication source~~

~~This means that the procedures no longer attempt to configure privilege separation with GSS Proxy. If needed, we can add steps to that purpose later after we agree on the necessary minimum proposed in this PR.~~

### Why are you introducing these changes? (Explanation, links to references, issues, etc.)

https://issues.redhat.com/browse/SAT-22855 requests making changes to how the existing procedure on direct AD integration + configuring AD as an external authentication source handles privilege separation for Apache with GSS proxy. While investigating that scenario, it turned out that proper privilege separation may currently not be possible to achieve. Therefore, the agreed-upon approach is to first document the basic steps for AD enrollment + authentication source configuration and then, in another PR, continue the work by adding steps to achieve proper privilege separation.

~~This PR originated with https://issues.redhat.com/browse/SAT-22855 and then... escalated.~~

~~According to https://issues.redhat.com/browse/SAT-22855, the procedure for direct AD integration with GSS-proxy needs updating. This PR continues the effort from https://github.com/theforeman/foreman-documentation/pull/2737.~~

~~Additionally, https://issues.redhat.com/browse/SAT-26355 requests adding two oddjob* packages to the same assembly.~~

### Anything else to add? (Considerations, potential downsides, alternative solutions you have explored, etc.)

~~Some might object to leaving Samba out of the picture. These resources explain why SSSD is preferred over Samba for direct AD integration:~~

~~* [1.1. Overview of direct integration using SSSD](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html-single/integrating_rhel_systems_directly_with_windows_active_directory/index#overview-of-direct-integration-using-sssd_connecting-rhel-systems-directly-to-ad-using-sssd)~~
~~* [6.1. Direct integration of Linux systems into Active Directory](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html-single/planning_identity_management/index#direct-integration-of-linux-systems-into-active-directory_planning-integration-with-ad)~~

~~I suspect that the procedures might have originally involved Samba because SSSD was not sufficiently mature when the procedures were written. However, SSSD has evolved in the past few years and can stand on its own now, except for limited edge cases.~~

### Checklists

* [x] I am okay with my commits getting squashed when you merge this PR.
* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [x] Foreman 3.11/Katello 4.13
* [x] Foreman 3.10/Katello 4.12
* [x] Foreman 3.9/Katello 4.11 (Satellite 6.15; orcharhino 6.8/6.9)
* [x] Foreman 3.8/Katello 4.10
* [x] Foreman 3.7/Katello 4.9 (Satellite 6.14)
* [x] Foreman 3.6/Katello 4.8
* [x] Foreman 3.5/Katello 4.7 (Satellite 6.13; orcharhino 6.6/6.7)
* We do not accept PRs for Foreman older than 3.5.

asteflova commented on guides/common/modules/proc_enrolling-server-with-the-ad-server.adoc:
```diff
@@ -18,7 +18,7 @@ Installing GSS-proxy and {nfs-client-package}:
 [options="nowrap", subs="+quotes,verbatim,attributes"]
 ----
-# {project-package-install} sssd adcli realmd ipa-python-compat krb5-workstation samba-common-tools
+# {project-package-install} adcli ipa-python-compat krb5-workstation oddjob oddjob-mkhomedir realmd samba-common-tools sssd
```
Confirmed in https://issues.redhat.com/browse/SAT-26355 and in [RHEL AD integration docs](https://issues.redhat.com/browse/SAT-26355).

asteflova replied:
Also requested in SAT-22855.

asteflova commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -59,6 +65,25 @@ euid = __ID_of_Apache_User__
 # chown root.{apache-user} /etc/httpd/conf/http.keytab
 # chmod 640 /etc/httpd/conf/http.keytab
 ----
+. Configure SSSD to map Group Policy Objects (GPOs) from AD to the `foreman` PAM service:
+.. Add the following lines to your `/etc/sssd/sssd.conf`:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+access_provider = ad
+ad_gpo_access_control = enforcing
+ad_gpo_map_service = +foreman
+----
+ifdef::satellite[]
++
+For more information on GPOs, see https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] i ...
+endif::[]
+.. Restart the `sssd` service:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# systemctl restart sssd
+----
```
Moved from the now deleted WARNING box for better visibility. Also, added a step to restart SSSD.

asteflova commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -93,25 +118,8 @@ Environment=GSS_USE_PROXY=1
-For more information on GPOs, see {RHELDocsBaseURL}9/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#how-sssd-interprets-gpo-access-control-rules_applying-group-policy-object-access-control-in-rhel[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}9 Integrating RHEL systems d ...
-endif::[]
-====
+. Add a *Service Principal Name (SPN)* in your AD server to enable Kerberos connections.
+For more information, see link:https://learn.microsoft.com/en-us/sql/database-engine/configure-windows/register-a-service-principal-name-for-kerberos-connections?view=sql-server-ver16[Register a Service Principal Name for Kerberos connections].
```
Requested in SAT-22855.

asteflova commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -51,6 +51,12 @@ cred_store = keytab:/etc/httpd/conf/http.keytab
 cred_store = ccache:/var/lib/gssproxy/clients/krb5cc_%U
 euid = __ID_of_Apache_User__
 ----
+. Add a SELinux context rule to apply the appropriate security context to the `http.keytab` file:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# semanage fcontext -a -s system_u -t krb5_keytab_t '/etc/httpd/conf/http\.keytab'
+----
```
Requested in SAT-22855.

asteflova reviewed (commented):
Because I'm not the original author of these changes, I'm adding a few notes that I made while familiarizing myself with the bug report. Feel free to ignore them.

Lennonka commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -59,6 +65,25 @@ euid = __ID_of_Apache_User__
+----
+ifdef::satellite[]
++
+For more information on GPOs, see https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] i ...
```
```suggestion
For more information on GPOs, see {RHELDocsBaseURL}8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] in _Integrating RHEL systems directly with Windows Active Directory_.
```
Also, why remove the link to RHEL 9?

asteflova replied:
Good catch, I didn't notice that happened. Re-adding the link.

Lennonka commented on guides/common/modules/proc_enrolling-server-with-the-ad-server.adoc:
```diff
@@ -18,7 +18,7 @@ Installing GSS-proxy and {nfs-client-package}:
 [options="nowrap", subs="+quotes,verbatim,attributes"]
 ----
-# {project-package-install} sssd adcli realmd ipa-python-compat krb5-workstation samba-common-tools
+# {project-package-install} adcli ipa-python-compat krb5-workstation oddjob oddjob-mkhomedir realmd samba-common-tools sssd
```
```suggestion
# {project-package-install} adcli ipa-python-compat krb5-workstation \
oddjob oddjob-mkhomedir realmd samba-common-tools sssd
```

Lennonka commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -59,6 +65,25 @@ euid = __ID_of_Apache_User__
+----
+ifdef::satellite[]
++
+For more information on GPOs, see link:{RHELDocsBaseURL}9/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#how-sssd-interprets-gpo-access-control-rules_applying-group-policy-object-access-control-in-rhel[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}9 Integrating RHEL syst ...
```
```suggestion
For more information on GPOs, see link:{RHELDocsBaseURL}9/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#how-sssd-interprets-gpo-access-control-rules_applying-group-policy-object-access-control-in-rhel[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}9 Integrating RHEL systems directly with Windows Active Directory_ or link:{RHELDocsBaseURL}8/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#applying-group-policy-object-access-control-in-rhel_managing-direct-connections-to-ad[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}8 Integrating RHEL systems directly with Windows Active Directory_.
```
:coffee: :)

lhellebr commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -51,6 +51,12 @@ cred_store = keytab:/etc/httpd/conf/http.keytab
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# semanage fcontext -a -s system_u -t krb5_keytab_t '/etc/httpd/conf/http\.keytab'
```
There was a suggestion to move the keytab to /etc/gssproxy, why keep it in /etc/httpd/conf when the purpose of this setup is to " implement stricter privilege separation for the Apache server by removing access to the keytab file while preserving Kerberos authentication functionality"?

lhellebr replied:
IMHO both locations (or any other) are good as long as correct permissions and SELinux context are set. I think with `/etc/gssproxy`, the semanage and restorecon commands may not be necessary.
Why not hold on to our own knowledge base though? https://access.redhat.com/articles/5854761

asteflova replied:
Ok, let's stick with /etc/gssproxy to match the KBase article. I haven't dropped semanage because according to https://issues.redhat.com/browse/SAT-22855, it should be necessary. But I can drop it if you tell me to @lhellebr

I think I can resolve this thread for now.

lhellebr commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -59,6 +65,25 @@ euid = __ID_of_Apache_User__
 # chown root.{apache-user} /etc/httpd/conf/http.keytab
 # chmod 640 /etc/httpd/conf/http.keytab
 ----
+. Configure SSSD to map Group Policy Objects (GPOs) from AD to the `foreman` PAM service:
```
Finally someone moved it from that confusing box that everyone ignored. Thanks!

lhellebr commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -93,25 +118,8 @@ Environment=GSS_USE_PROXY=1
-For more information on GPOs, see {RHELDocsBaseURL}9/html/integrating_rhel_systems_directly_with_windows_active_directory/managing-direct-connections-to-ad_integrating-rhel-systems-directly-with-active-directory#how-sssd-interprets-gpo-access-control-rules_applying-group-policy-object-access-control-in-rhel[How SSSD interprets GPO access control rules] in _{RHEL}{nbsp}9 Integrating RHEL systems d ...
-endif::[]
-====
+. Add a *Service Principal Name (SPN)* in your AD server to enable Kerberos connections.
```
I've never had to do this since this command (from docs 5.3.3. Configuring Direct AD Integration with GSS-Proxy, step 6) adds Satellite as a client service - that is, adds it to AD and gets its keytab:
```
KRB5_KTNAME=FILE:/etc/httpd/conf/http.keytab net ads keytab add HTTP -U administrator -d3 -s /etc/net-keytab.conf
```

lhellebr replied:
Of course use of `administrator` may be questionable...

lhellebr commented on guides/common/modules/proc_enrolling-server-with-the-ad-server.adoc:
```diff
@@ -18,7 +18,8 @@ Installing GSS-proxy and {nfs-client-package}:
 ----
-# {project-package-install} sssd adcli realmd ipa-python-compat krb5-workstation samba-common-tools
+# {project-package-install} adcli ipa-python-compat krb5-workstation \
+oddjob oddjob-mkhomedir realmd samba-common-tools sssd
```
I've never had to specify `oddjob*` to get the setup working

lhellebr replied:
So this is what the `realm join` command actually does when I don't install `oddjob*` before running it:
```
 * Couldn't find file: /usr/sbin/oddjobd
 * Required files: /usr/sbin/oddjobd, /usr/libexec/oddjob/mkhomedir, /usr/sbin/sssd, /usr/bin/net
 * Resolving required packages
 * Installing necessary packages: oddjob oddjob-mkhomedir samba-common-tools
```
Can't hurt to pre-install it manually, not relying on the above to work. Also, it's the documented way. So yes, let's keep it.

lhellebr commented:
I've commented on some issues based on my memory and I will proceed with testing this actual workflow documented, verbatim, without changes I've suggested.

asteflova commented:
I believe I have resolved all comments you've posted so far @lhellebr Let me know what else needs changing. Thanks!

lhellebr commented:
Tested with the latest version and when running the installer, I get the following error.

```
2024-07-24 12:18:57 [NOTICE] [configure] System configuration has finished.

Error 1: Puppet Exec resource 'ipa-getkeytab' failed. Logs:
  /Stage[main]/Foreman::Config/Exec[ipa-getkeytab]/before
    before to File[/etc/httpd/conf/http.keytab]
  /Stage[main]/Foreman::Config/Exec[ipa-getkeytab]
    Starting to evaluate the resource (273 of 1654)
    Failed to call refresh: '/bin/echo Get keytab           && KRB5CCNAME=KEYRING:session:get-http-service-keytab kinit -k           && KRB5CCNAME=KEYRING:session:get-http-service-keytab /usr/sbin/ipa-getkeytab -k /etc/httpd/conf/http.keytab -p HTTP/<FQDN>           && kdestroy -c KEYRING:session:get-http-service-keytab' returned 1 instead of one of [0]
    '/bin/echo Get keytab           && KRB5CCNAME=KEYRING:session:get-http-service-keytab kinit -k           && KRB5CCNAME=KEYRING:session:get-http-service-keytab /usr/sbin/ipa-getkeytab -k /etc/httpd/conf/http.keytab -p HTTP/<FQDN>           && kdestroy -c KEYRING:session:get-http-service-keytab' returned 1 instead of one of [0]
    Evaluated in 0.04 seconds
  /Stage[main]/Foreman::Config/Exec[ipa-getkeytab]/creates
    Checking that 'creates' path '/etc/httpd/conf/http.keytab' exists
    Checking that 'creates' path '/etc/httpd/conf/http.keytab' exists
  Exec[ipa-getkeytab](provider=posix)
    Executing '/bin/echo Get keytab           && KRB5CCNAME=KEYRING:session:get-http-service-keytab kinit -k           && KRB5CCNAME=KEYRING:session:get-http-service-keytab /usr/sbin/ipa-getkeytab -k /etc/httpd/conf/http.keytab -p HTTP/<FQDN>           && kdestroy -c KEYRING:session:get-http-service-keytab'
    Executing '/bin/echo Get keytab           && KRB5CCNAME=KEYRING:session:get-http-service-keytab kinit -k           && KRB5CCNAME=KEYRING:session:get-http-service-keytab /usr/sbin/ipa-getkeytab -k /etc/httpd/conf/http.keytab -p HTTP/<FQDN>           && kdestroy -c KEYRING:session:get-http-service-keytab'
  /Stage[main]/Foreman::Config/Exec[ipa-getkeytab]/returns
    Get keytab
    kinit: Cannot determine realm for host (principal host/<FQDN>@)
    change from 'notrun' to ['0'] failed: '/bin/echo Get keytab           && KRB5CCNAME=KEYRING:session:get-http-service-keytab kinit -k           && KRB5CCNAME=KEYRING:session:get-http-service-keytab /usr/sbin/ipa-getkeytab -k /etc/httpd/conf/http.keytab -p HTTP/<FQDN>           && kdestroy -c KEYRING:session:get-http-service-keytab' returned 1 instead of one of [0]
    Get keytab
    kinit: Cannot determine realm for host (principal host/<FQDN>@)

1 error was detected during installation.
Please address the errors and re-run the installer to ensure the system is properly configured.
Failing to do so is likely to result in broken functionality.
```
Note that the realm (In this log, `<FQDN>` is doctored to make it public) is derived from the host's FQDN which is different to the realm. Normally, the realm is taken from the config files created as per docs but here, the installer doesn't use it.
My hypothesis is it's because the installer expects the keytab to be in a certain place (that changed) but I can't confirm it yet.

Additionally, in part 5.8.3 step 8a, it's unclear which section of the file the text should be added to.

asteflova commented:
> My hypothesis is it's because the installer expects the keytab to be in a certain place (that changed) but I can't confirm it yet.

I don't think I can help with that but let me know if I'm missing something and there is anything I'm expected to do about this.

> Additionally, in part 5.8.3 step 8a, it's unclear which section of the file the text should be added to.

I clarified that the lines go to the domain section for the AD domain. And I also added a reference to the sssd-ad man page (there are multiple man pages related to sssd.conf so it might be helpful to point users to the most relevant one).

lhellebr commented:
>>My hypothesis is it's because the installer expects the keytab to be in a certain place (that changed) but I can't confirm it yet.
>
>I don't think I can help with that but let me know if I'm missing something and there is anything I'm expected to do about this.

It means what we agreed to do here https://github.com/theforeman/foreman-documentation/pull/3149#discussion_r1688115759 and what was suggested in the ticket was kinda illegal. Not by the nature of the change, after all, we followed the KB... but it seems the satellite EXPECTS the keytab to be in `/etc/httpd/conf/http.keytab`. It can be fixed either by:
1) Devels changing this expectation. This would be justifiable but IMO not necessary. CC @adamruzicka 
2) You just changing it back to the original path. Effectively negating what is perhaps the main point of the whole ticket - but solving the same problem in a different manner.

Anyway, this must be fixed, it's a blocker, the process doesn't work as is.

lhellebr commented:
When exactly following the docs in this PR's current state but changing `/etc/gssapi/http.conf` to `/etc/httpd/conf/http.keytab`, I got a working configuration.

adamruzicka commented:
> but solving the same problem in a different manner.

So if we kept the old path (`/etc/httpd/conf/http.keytab`), then this would reduce the changes here to:
- setting a different selinux label on it
- setting a different user group on it

If this is enough to resolve the issue then I'd prefer it over changing the puppet modules to expect the keytab to be in a different place, especially if we want to have this done soon.

ekohl commented:
> If this is enough to resolve the issue then I'd prefer it over changing the puppet modules to expect the keytab to be in a different place, especially if we want to have this done soon.

You can use `--foreman-http-keytab /path/to/keytab`. It just defaults to `${apache::conf_dir}/http.keytab`.

I'm not sure about the SELinux policy, because that comes from the base SELinux policy. Looks like the default is `/etc/httpd/conf/keytab` which we never used.

```console
# semanage fcontext -l | grep keytab
/etc/httpd/conf/keytab                             regular file       system_u:object_r:httpd_keytab_t:s0 
/etc/krb5\.keytab                                  all files          system_u:object_r:krb5_keytab_t:s0 
/etc/krb5kdc/kadm5\.keytab                         regular file       system_u:object_r:krb5_keytab_t:s0 
/var/kerberos/krb5(/.*)?                           all files          system_u:object_r:krb5_keytab_t:s0 
/var/kerberos/krb5kdc/kadm5\.keytab                regular file       system_u:object_r:krb5_keytab_t:s0 
```

Looking at the procedure it states `The Apache user must not have access to the keytab file.` so then it is indeed very illogical to use the Apache config directory.

ekohl commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -40,24 +40,50 @@ security = ads
+----
+# systemctl restart sssd
 ----
 . Enable IPA authentication in {Project}:
```
Looking at this, I doubt it's really correct. That feature is really designed for IPA, not for AD.

For example, it ensures the keytab is owned by Apache, which directly contradicts the statement that Apache shouldn't be able to read the keytab. However, it does set up some things in Apache that may be needed.

Big question: what is actually needed for the AD integration?

lhellebr replied:
I believe in this case, AD is masked as IPA by SSSD. That's how I always imagined it to work.

ekohl replied:
Further elaborating on this: we explicitly tell Apache about its location: https://github.com/theforeman/puppet-foreman/blob/5f38479c870ef03a4a1565007be68b61e8950d53/templates/auth_gssapi.conf.erb#L6

ekohl replied:
It also inspired me to refactor a bit: https://github.com/theforeman/puppet-foreman/pull/1172

ekohl replied:
> I believe in this case, AD is masked as IPA by SSSD. That's how I always imagined it to work.

I'd classify it as a hack and fundamentally there are some differences here that IMHO make it incompatible. In particular, the keytab handling is different. At least when I read this documentation and the code for IPA.

lhellebr replied:
Unaccessibility by Apache can still be achieved using permissions and selinux context on that file, right?

lhellebr replied:
Does Apache only need to read it during installation or does it access it regularly? If the latter, we either need to change the code or the goal outlined in the docs can't be reached IMO.

lhellebr replied:
Anyway, when I used path `/etc/httpd/config/http.conf` with `root.root`, `600` and `krb5_keytab_t`, it worked for me.

lhellebr replied:
```
# sealert -l 20295c58-da31-43c5-897c-63c5e83544e0
SELinux is preventing /usr/sbin/gssproxy from search access on the directory httpd.

*****  Plugin catchall (100. confidence) suggests   **************************

If you believe that gssproxy should be allowed search access on the httpd directory by default.
Then you should report this as a bug.
You can generate a local policy module to allow this access.
Do
allow this access for now by executing:
# ausearch -c 'gssproxy' --raw | audit2allow -M my-gssproxy
# semodule -X 300 -i my-gssproxy.pp


Additional Information:
Source Context                system_u:system_r:gssproxy_t:s0
Target Context                system_u:object_r:httpd_config_t:s0
Target Objects                httpd [ dir ]
Source                        gssproxy
Source Path                   /usr/sbin/gssproxy
Port                          <Unknown>
Host                         <FQDN>
Source RPM Packages           
Target RPM Packages           
SELinux Policy RPM            selinux-policy-targeted-3.14.3-139.el8_10.noarch
Local Policy RPM              selinux-policy-targeted-3.14.3-139.el8_10.noarch
Selinux Enabled               True
Policy Type                   targeted
Enforcing Mode                Enforcing
Host Name                     <FQDN>
Platform                      Linux
                              <FQDN>
                              4.18.0-553.8.1.el8_10.x86_64 #1 SMP Fri Jun 14
                              03:19:37 EDT 2024 x86_64 x86_64
Alert Count                   9
First Seen                    2024-07-25 05:10:52 EDT
Last Seen                     2024-07-25 09:12:22 EDT
Local ID                      20295c58-da31-43c5-897c-63c5e83544e0

Raw Audit Messages
type=AVC msg=audit(1721913142.873:1466): avc:  denied  { search } for  pid=27259 comm="gssproxy" name="httpd" dev="vda3" ino=159424064 scontext=system_u:system_r:gssproxy_t:s0 tcontext=system_u:object_r:httpd_config_t:s0 tclass=dir permissive=0


Hash: gssproxy,gssproxy_t,httpd_config_t,dir,search
```

lhellebr replied:
And the autogenerated rule is:
```
module my-gssproxy 1.0;

require {
        type httpd_config_t;
        type gssproxy_t;
        class dir search;
}

#============= gssproxy_t ==============
allow gssproxy_t httpd_config_t:dir search;
```
That looks like gssproxy just isn't supposed to look into Apache's folder. It all looks hacky.
Surprisingly, it works.

asteflova replied:
With the latest commit, I went with @lhellebr's "hacky but works" approach.

ekohl commented on guides/common/modules/proc_enrolling-server-with-the-ad-server.adoc:
```diff
@@ -18,7 +18,8 @@ Installing GSS-proxy and {nfs-client-package}:
 [options="nowrap", subs="+quotes,verbatim,attributes"]
 ----
-# {project-package-install} sssd adcli realmd ipa-python-compat krb5-workstation samba-common-tools
+# {project-package-install} adcli ipa-python-compat krb5-workstation \
```
I wonder why we have prerequisites in this procedure to install some packages, but then install some more packages. Shouldn't that be a single command?

Also, why do we need `nfs-utils` for this? It was added in 9c4054406c9076db5abb2447dd09f4d4b8f135a6 but conceptually I don't see why it would be needed.

Also, on EL9 there is no `ipa-python-compat` package.

asteflova replied:
You make a good point. If possible, I'd like to drop this for now until we review the whole user story.

asteflova commented:
If possible, I'd like to limit this PR to addressing https://issues.redhat.com/browse/SAT-22855 in a way that enables us to close that ticket. That would involve:
* Reviewing the keytab's location and permissions (ongoing conversation)
* A decision on whether to include a step for SPN in AD (resolved)
* Adding oddjob* packages (resolved)

The last time I tried to look at the whole AD/SSSD integration procedure from a bigger perspective, it resulted in https://github.com/theforeman/foreman-documentation/pull/3056 which became unmanageable and I had to close it. Because it's absolutely true that there are larger issues to deal with. Can we track them in https://issues.redhat.com/browse/SAT-26039 in order to avoid blocking this PR?

lhellebr commented:
There was a discussion about that and it may happen in the future, not in this PR though. Also note that scenario is not currently tested or automated.

asteflova commented:
> @asteflova you mentioned that it's more common to deploy IPA and then link your IPA and AD installations.

I'm afraid I didn't explain myself correctly at the meeting. Connecting directly to AD (without using IPA) is a valid use case. For example, you want direct integration if you are only connecting a small number of systems to AD.

It's okay that [5.10. Configuring Active Directory as an external identity provider for Foreman (link to nightly)](https://docs.theforeman.org/nightly/Installing_Server/index-katello.html#configuring-active-directory-as-an-external-identity-provider-for-project_foreman) wants to document it. The highly suspicious thing I mentioned at the meeting is that [5.10.5. Configuring the FreeIPA server to use cross-forest trust (nightly)](https://docs.theforeman.org/nightly/Installing_Server/index-katello.html#Configuring_the_FreeIPA_Server_to_use_Cross_Forest_Trust_foreman) talks about configuring an IPA server as part of that use case. If your goal is direct integration (without IPA), then you shouldn't need to use IPA (but 5.10.5. asks you to do just that).

Per the [RHEL Windows integration guide](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/planning_identity_management/planning-integration-with-ad_planning-identity-management#direct-integration-of-linux-systems-into-active-directory_planning-integration-with-ad):

- 6.1. Direct integration of Linux systems into Active Directory: "In direct integration, Linux systems are connected directly to Active Directory (AD)."

Compare that to:

- 6.2. Indirect integration of Linux systems into Active Directory by using Identity Management: "In indirect integration, Linux systems are first connected to a central server which is then connected to Active Directory (AD)." That central server is IPA.

asteflova commented:
> The highly suspicious thing I mentioned at the meeting is that [5.10.5. Configuring the FreeIPA server to use cross-forest trust (nightly)](https://docs.theforeman.org/nightly/Installing_Server/index-katello.html#Configuring_the_FreeIPA_Server_to_use_Cross_Forest_Trust_foreman) talks about configuring an IPA server as part of that use case.

... and I think I've just realized why the procedure is there. Spoiler alert: It's kind of my fault and I'll send a PR to fix it soon.

ekohl commented:
> ... and I think I've just realized why the procedure is there. Spoiler alert: It's kind of my fault and I'll send a PR to fix it soon.

![image](https://github.com/user-attachments/assets/32627a16-3057-44b6-815b-c7034f9f6c9c)

asteflova commented:
Well it's a good thing that there are other things wrong with the procedures too! (I'd feel pretty awkward if there weren't.)

I'll split out the non-controversial parts of this PR into other PRs to get those merged at least. For the trickier parts concerning the whole feature, let's discuss them with Endeavour and Platform first to try to figure out a solution.

asteflova commented:
@ekohl @lhellebr I looked into dropping the GSS proxy parts of the user story ([5.8. Configuring Active Directory as an external identity provider for Satellite](https://theforeman-foreman-documentation-preview-pr-3149.surge.sh/nightly/Installing_Server/index-satellite.html#configuring-active-directory-as-an-external-identity-provider-for-project_satellite)) and didn't find a way to do that. The GSS proxy parts seem intertwined with the rest too much.

As an alternative, I propose to merge the following changes. These are the only changes requested in https://issues.redhat.com/browse/SAT-22855 that I think we are able to implement now and [Lukáš got it working](https://github.com/theforeman/foreman-documentation/pull/3149/files#r1691173944) this way:

* The keytab is stored back in its original location (/etc/httpd/conf/htt.keytab)
* SELinux context is applied to the keytab file, ownership set to root.root and permissions to 600
* oddjob* packages added

As a reminder, this is only temporary. I absolutely want to get to the bottom of this but it will take some time, a cross-team effort, and a lot of ice cream. In the meantime, we can at least get this partial fix merged for now. What do you think?

ekohl commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -51,13 +51,39 @@ cred_store = keytab:/etc/httpd/conf/http.keytab
 # KRB5_KTNAME=FILE:/etc/httpd/conf/http.keytab net ads keytab add HTTP -U administrator -d3 -s /etc/net-keytab.conf
-# chown root.{apache-user} /etc/httpd/conf/http.keytab
-# chmod 640 /etc/httpd/conf/http.keytab
+# chown root.root /etc/httpd/conf/http.keytab
+# chmod 600 /etc/httpd/conf/http.keytab
```
The installer enforces different permissions
```suggestion
```

ekohl commented on guides/common/modules/proc_configuring-direct-ad-integration-with-gss-proxy.adoc:
```diff
@@ -40,7 +40,7 @@ security = ads
 ----
 +
-Apache user must not have access to the keytab file.
+The Apache user must not have access to the keytab file.
```
The installer enforces that the Apache user does have access
```suggestion
```

ekohl reviewed (changes requested):
As I tried to explain, the use of gssproxy and `{foreman-installer} --foreman-ipa-authentication=true` does not make sense in the current implementation. Fundamentally, the installer will enforce that Apache **can** read the keytab and the whole purpose of gssproxy is to ensure Apache **can't** read the keytab and still work.

This is why I suggested to simplify the whole procedure to drop gssproxy. We can backport that to existing support versions. Then in the next version (timing wise, probably Foreman 3.13) we can implement support for gssproxy.

asteflova commented:
> As I tried to explain, the use of gssproxy and `{foreman-installer} --foreman-ipa-authentication=true` does not make sense in the current implementation. Fundamentally, the installer will enforce that Apache **can** read the keytab and the whole purpose of gssproxy is to ensure Apache **can't** read the keytab and still work.
> 
> This is why I suggested to simplify the whole procedure to drop gssproxy. We can backport that to existing support versions. Then in the next version (timing wise, probably Foreman 3.13) we can implement support for gssproxy.

I heard you and I did try dropping the GSS proxy elements of the current procedures. I've updated this PR to show what I ended up with: a procedure that misses several steps and no way to test it. I didn't feel confident enough to go in that direction.

On the other hand, the alternative was to keep the GSS proxy parts included even though the whole configuration doesn't exactly do what GSS proxy promises to be. This was confirmed to result in a working setup: https://github.com/theforeman/foreman-documentation/pull/3149#issuecomment-2249893886  which is why it seemed like a better choice to me.

A third alternative could be to just close this PR, focus on finding a proper way to configure direct AD integration, and accept that we won't try to find an interim solution (which is what this PR has turned into). This would be my preference. What are your thoughts @ekohl?

asteflova commented:
Rebased and split off two changes into separate PRs: https://github.com/theforeman/foreman-documentation/pull/3189 and https://github.com/theforeman/foreman-documentation/pull/3190

Lennonka reviewed (approved):
LGTM style wise

asteflova commented:
Switching back to draft because this is nowhere near ready for merging. Also, updating the PR title to reflect the change of scope. My plan is to get my hands on an AD machine and try to connect a Foreman server to the AD realm by following the RHEL docs, more specifically using the SSSD-based procedure because that’s the recommended one. 

The assumption I’m making is that we should be able to treat the Foreman server base system as any other RHEL machine for the actual join (thus following the standard procedure documented on the RHEL side) and then run additional steps, like --foreman-ipa-authentication=true, on top of that. That would be preferable to the current situation where we are basically providing our own version of the AD join procedure.

Once we have a basic procedure like that in place, we will be in a better position to think about bringing the GSS Proxy keytab separation into the picture.

asteflova commented:
After talking to an SSSD engineer, I've determined the bare minimum that should result in the Foreman server recognizing AD as an authentication source. Emphasis on "should".

However, approaching the procedure like this -- referencing RHEL docs as much as possible -- would help us out of the current situation where on the Foreman side, we are basically documenting our own way of joining an AD domain that diverges from what the SSSD team actually recommends and maintains.

I still don't have an environment to test the workflow but I'm getting closer to obtaining one.

asteflova commented:
Thanks to @adamruzicka, we now have a working procedure to join a Foreman server directly to an AD domain and configure the AD server as an external authentication source.

asteflova commented:
I'm marking the procedure as ready for review to open it up for comments, but I will also reach out to a couple of people for more detailed review regarding esp. QE testing and Samba.

asteflova commented:
I also updated the PR description in case any potential reviewers need a recap.

maximiliankolb commented on guides/common/modules/proc_configuring-direct-ad-integration.adoc:
file name does not match anchor/title. Please adjust.

asteflova replied:
Fixed.

maximiliankolb commented on guides/common/modules/proc_configuring-direct-ad-integration.adoc:
```diff
@@ -0,0 +1,56 @@
+
+.Verification
+* Log in to {ProjectwebUI} by entering the credentials of a user defined in AD.
+Enter the user name and the domain name, for example: _ad_user_@_ad-example.com_.
```
```suggestion
Enter the user name and the domain name, for example: _ad_user_@_ad.example.com_.
```

maximiliankolb replied:
The example URL differs to https://github.com/theforeman/foreman-documentation/pull/3149/files#diff-fb730f56da8be9208251e9209ac1344d1007e6c04e30a89a6ea82e2ac0b3d708R30

asteflova replied:
I unified all references to the example AD realm and user.

maximiliankolb commented on guides/common/modules/proc_configuring-direct-ad-integration.adoc:
```diff
@@ -0,0 +1,56 @@
+Enter the user name and the domain name, for example: _ad_user_@_ad-example.com_.
+
+.Additional resources
+
```
```suggestion
```

asteflova replied:
Fixed.

adamruzicka commented on guides/common/modules/proc_configuring-direct-ad-integration.adoc:
```diff
@@ -0,0 +1,56 @@
+----
+# systemctl restart sssd
+----
+. Configure AD as the authentication source for your {ProjectServer}:
+.. Create an Apache keytab file if it does not already exist on your system:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# touch /etc/httpd/conf/http.keytab
+----
++
+You can leave the file empty.
+The `{foreman-installer}` command fails to configure the authentication source if the file is missing.
```
To be clear, this is a workaround. The installer deploys things to use gssproxy and tries to obtain a keytab in IPA-specific ways, which fails in a sssd-only, IPA-less scenario. Making gssproxy use optional (something along the lines of [1] + migrations in installer) could help in not having to carry the workaround, if we want to have this as a first class scenario.

cc @ekohl 

[1] - https://github.com/theforeman/puppet-foreman/compare/master...adamruzicka:optional-gssproxy?expand=1

adamruzicka commented on guides/common/modules/proc_configuring-direct-ad-integration.adoc:
```diff
@@ -0,0 +1,56 @@
+----
+# {foreman-installer} --foreman-ipa-authentication=true
+----
+.. Restart Apache:
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# systemctl restart httpd
+----
```
Again, a workaround for an installer issue

adamruzicka commented on guides/common/modules/proc_enrolling-server-with-the-ad-server.adoc:
```diff
@@ -1,35 +1,32 @@
 [options="nowrap", subs="+quotes,verbatim,attributes"]
 ----
-# {project-package-install} gssproxy {nfs-client-package}
+# {foreman-maintain} samba-common-tools realmd oddjob oddjob-mkhomedir sssd adcli krb5-workstation
```
What does `{foreman-maintain}` expand to?

asteflova replied:
Fixed.

adamruzicka commented on guides/common/modules/proc_configuring-direct-ad-integration.adoc:
```diff
@@ -0,0 +1,56 @@
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+[domain/_EXAMPLE_]
+access_provider = ad
```
This line seems to be already there after doing `realm join`

asteflova replied:
Removed.

adamruzicka reviewed (commented):
Mostly context for reviewers

asteflova commented on guides/common/modules/proc_enrolling-server-with-the-ad-server.adoc:
```diff
@@ -1,35 +1,32 @@
-* GSS-proxy and {nfs-client-package} are installed.
-+
-Installing GSS-proxy and {nfs-client-package}:
+For more information, see link:https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/8/html-single/integrating_rhel_systems_directly_with_windows_active_directory/index#connecting-rhel-systems-directly-to-ad-using-sssd_integrating-rhel-systems-directly-with-active-directory[Connecting RHEL systems directly to AD using SSSD].
```
Actually, let's try to drop the whole module and mention the requirement for the Foreman server to be a member of the AD domain as a prerequisite in the next section. I am keeping the link for all builds because SSSD doesn't have comprehensive user documentation upstream.

The whole workflow looks much simpler now, which is what we want.

mmuehlfeldRH commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,58 @@
++
+You can leave the file empty.
+The `{foreman-installer}` command fails to configure the authentication source if the file is missing.
+.. Enable IPA authentication:
```
It's the first time that I read "IPA" in this section.

* Should it be expanded?
* Section 5.4 mentions "Identity Management" as an alternative to direct AD integration. Why now IPA? If it doesn't matter in the background if IPA or AD, can you rephrase the text?

asteflova replied:
This is tricky. I kept "IPA" literal and un-expanded on purpose. "Identity Management" is not part of this scenario so I can't use that. The `--foreman-ipa-authentication` option is used when connecting to an IdM identity store (indirect integration, as described in section 5.4), but also for the SSSD/AD scenarios (direct integration, as described in this PR). I assume that's because in both indirect and direct scenarios, `--foreman-ipa-authentication` actually configures an SSSD client. So in a sense, when I decided to go with "enable IPA authentication", I just translated the `--foreman-ipa-authentication` option's name.

I think I can go with something generic like "Enable the authentication source". You are right that talking about "IPA" would raise too many questions.

mmuehlfeldRH commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,58 @@
+
+.Verification
+* Log in to {ProjectwebUI} by entering the credentials of a user defined in AD.
+Enter the user name and the domain name, for example: _ad_user_@_ad.example.com_.
```
If there are no separate fields for user name and domain name, then I recommend:

Enter the user name in user principal name (UPN) format, for example _ad_user_@_ad.example.com_.

mmuehlfeldRH replied:
You can mention here that this requires some specific browser settings.

With respect to "[5.4.2. Kerberos configuration in web browsers](https://theforeman-foreman-documentation-preview-pr-3149.surge.sh/nightly/Installing_Server/index-satellite.html#Kerberos_Configuration_in_Web_Browsers_satellite)": [IE is EOL](https://learn.microsoft.com/en-us/lifecycle/products/internet-explorer-11). I would mention Firefox, Chrome, an Edge.

asteflova replied:
> If there are no separate fields for user name and domain name, then I recommend:
> 
> Enter the user name in user principal name (UPN) format, for example _ad_user_@_ad.example.com_.

This is great, I wasn't aware of the term UPN. Thanks!

asteflova replied:
> You can mention here that this requires some specific browser settings.

You mean being able to log in with just the user name? That's already in the works, I hope to submit another PR once I finalize this one. (More changes will be needed to cover that and I don't want this PR to grow out of scope... again.)

> With respect to "[5.4.2. Kerberos configuration in web browsers](https://theforeman-foreman-documentation-preview-pr-3149.surge.sh/nightly/Installing_Server/index-satellite.html#Kerberos_Configuration_in_Web_Browsers_satellite)": [IE is EOL](https://learn.microsoft.com/en-us/lifecycle/products/internet-explorer-11). I would mention Firefox, Chrome, an Edge.

O-oh. I.. think we will need to take a look at the whole section then. If it's that old, we should review it more thoroughly. Thanks again!

mmuehlfeldRH commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,58 @@
+Enter the user name and the domain name, for example: _ad_user_@_ad.example.com_.
+
+.Additional resources
+* For more information about the options to configure an AD provider for SSSD, see the `sssd-ad(5)` man page.
```
To be compliant with the [RH SSG](https://redhat-documentation.github.io/supplementary-style-guide/#man-pages):
* `sssd-ad(5)` man page on your system

asteflova replied:
Good to know!

mmuehlfeldRH reviewed (commented):
I finished the peer review. Good work.

asteflova reviewed (commented):
Thanks a lot for the review @mmuehlfeldRH! This was very useful.

To introduce @mmuehlfeldRH, he is a very experienced tech writer with plenty of experience also in the area of Linux/AD integration. I believe that with his feedback, I can consider style review to be "done". But if anyone has any other concerns about style on this PR, more feedback is still welcome.

asteflova commented:
One thing I did not need to realize on a Friday afternoon is that the new steps do not configure Kerberos SSO for AD users.

```
# kinit ad_user
# curl -k -u : --negotiate https://foreman.example.com/users/extlogin
# <html><meta http-equiv="refresh" content="0; URL=/users/login"><body>Kerberos authentication did not pass.</body></html>
```

We will need to figure out what steps need to be taken to configure SSO, but perhaps a subsequent PR will do.

asteflova commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+
+Enable Active Directory (AD) users to access {Project} by configuring the corresponding authentication provider on your {ProjectServer}.
+
+.Prerequisites
```
The prerequisites section is unusually long. I don't like that but I still think it's better than turning it into a separate procedure module. It's still a single prerequisite ("the system must be enrolled into the AD domain under a certain configuration") and adding the two commands is designed to make sure the instructions are clear enough.

asteflova replied:
I wish I didn't have to specify the list of packages and the command to join AD explicitly like this (referring to RHEL docs would be the preferred choice because the RHEL IdM team is in sync with the engineering team developing these use cases) but unfortunately, Foreman requires a specific setup that combines SSSD and Samba. So the list of packages and the command to join the domain is different from what the RHEL docs recommend.

mmuehlfeldRH replied:
I don't like commands in prerequisites. In most cases, it's possible to move the commands to the beginning of the procedure or to a separate procedure module (and then link it).

asteflova commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+For more information on direct AD integration, see link:https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html-single/integrating_rhel_systems_directly_with_windows_active_directory/index#connecting-rhel-systems-directly-to-ad-using-samba-winbind_integrating-rhel-systems-directly-with-active-directory[Connecting RHEL systems directly to AD using Samba Winbind].
+
+.Procedure
+. Define AD realm configuration in a location where {foreman-installer} expects it:
```
I'm not sure about the wording here. I do think it's helpful to explain why we are asking users to create a directory named "IPA" when FreeIPA/Identity Management is not involved at all.

mmuehlfeldRH replied:
I have the same question. If SSSD (or other software) expect this file, it should be spelled out.

asteflova replied:
I believe it's the foreman-installer utility that does the expecting, so I hope we have this concern covered :) Thanks for confirming.

asteflova commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+----
+# {foreman-installer} --foreman-ipa-authentication=true
+----
+
```
I didn't need to restart httpd to be able to log in so I dropped that step.

asteflova commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+[global]
```
This previously included the line "server = unused". But the [option is deprecated](https://linux.die.net/man/5/default.conf) and the setup works for me without it.

asteflova commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# {foreman-maintain} packages install krb5-workstation oddjob-mkhomedir oddjob realmd samba-winbind-clients samba-winbind samba-common-tools samba-winbind-krb5-locator sssd
```
The list of packages previously included `adcli`. But the procedure uses `realm` instead. The setup worked for me without installing `adcli`.

asteflova commented:
I moved the SSSD-only procedure to https://github.com/theforeman/foreman-documentation/pull/3286 where I want to continue exploring that path. I tried using a different utility to generate the keytab (the currently documented one relies on Samba) but didn't succeed. I can't afford to block fixing this procedure with this, so with this PR, I'm returning to the SSSD+Samba-based solution. This is also known as reinventing the wheel or, as I prefer to call it, a 'valuable learning opportunity'. I also updated this PR's description.

maximiliankolb commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# {foreman-maintain} packages install krb5-workstation oddjob-mkhomedir oddjob realmd samba-winbind-clients samba-winbind samba-common-tools samba-winbind-krb5-locator sssd
```
```suggestion
# {project-package-install} krb5-workstation oddjob-mkhomedir oddjob realmd samba-winbind-clients samba-winbind samba-common-tools samba-winbind-krb5-locator sssd
```

See https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc#L192 and https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-satellite.adoc#L100

asteflova replied:
Thanks. And sorry! I keep forgetting about this.

maximiliankolb commented on guides/common/modules/con_configuring-active-directory-as-an-external-identity-provider-for-project.adoc:
```diff
@@ -1,23 +1,22 @@
 
-This section shows how to use direct Active Directory (AD) as an external authentication source for {ProjectServer}.
+If the base system of your {ProjectServer} is connected directly to Active Directory (AD), you can configure AD as an external authentication source for {Project}.
+Direct AD integration means that a Linux system is joined directly to the AD domain where the identity is stored. The following login methods are available for AD users:
```
```suggestion
Direct AD integration means that a Linux system is joined directly to the AD domain where the identity is stored.
The following login methods are available for AD users:
```

Linux system = host aka. registered to Foreman? Foreman itself? any host?

asteflova replied:
This is a general definition of Linux system. In this case, it's the base system of Foreman server. But I thought that since I was defining the term in general terms, I should use "Linux system" as a general term as well.

maximiliankolb commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+Enable Active Directory (AD) users to access {Project} by configuring the corresponding authentication provider on your {ProjectServer}.
+
+.Prerequisites
+* The base system of your {ProjectServer} must be joined to an Active Directory (AD) domain.
```
suggestions "Foreman Server has joined ..." or "You can connected Foreman server with ...". Your call.

asteflova replied:
Addressed in another comment.

maximiliankolb commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+
+.Prerequisites
+* The base system of your {ProjectServer} must be joined to an Active Directory (AD) domain.
+To enable AD users to sign in with Kerberos single sign-on, join the system by using the System Security Services Daemon (SSSD) and Samba services.
```
What does "the system" refer to? Considering that the command uses `foreman-maintain`, I assume it's either Foreman Server, or, more generic, Smart Proxies?

asteflova replied:
It refers to "the base system of your {ProjectServer}" from the previous line. This part really talks about connecting the underlying system of the Foreman server, not about connecting the Foreman server itself. (Because the part about connecting the Foreman server comes later with `foreman-installer --foreman-ipa-authentication=true`.) Do you think it works this way? Or would you still change the wording @maximiliankolb?

maximiliankolb replied:
@asteflova I still lean towards `{ProjectServer}` to emphasis that it's a task done on Foreman server, not your local host or any registered host. But I am also OK with "base system". Your call.

asteflova replied:
I reworded the line to help emphasize the point. I hope it helps, if not, we can always revisit :)

maximiliankolb commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+* The base system of your {ProjectServer} must be joined to an Active Directory (AD) domain.
+To enable AD users to sign in with Kerberos single sign-on, join the system by using the System Security Services Daemon (SSSD) and Samba services.
++
+Ensure you have the following packages installed:
```
```suggestion
Install the following packages on {ProjectServer}:
```

maximiliankolb commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+# realm join _AD.EXAMPLE.COM_ --membership-software=samba --client-software=sssd
+----
++
+For more information on direct AD integration, see link:https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html-single/integrating_rhel_systems_directly_with_windows_active_directory/index#connecting-rhel-systems-directly-to-ad-using-samba-winbind_integrating-rhel-systems-directly-with-active-directory[Connecting RHEL systems directly to AD using Samba Winbind].
```
There's an attribute for "docs.redhat.com": https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc#L27

maximiliankolb commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+$ curl -k -u : --negotiate https://{foreman-example-com}/users/extlogin
+<html><body>You are being <a href="{foreman-example-com}/hosts">redirected</a>.</body></html>
```
suggestion: split off the second line because it's stdout whereas curl is the command.

asteflova replied:
Much better, thanks.

maximiliankolb commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,121 @@
+----
+$ kinit _ad_user_@_AD.EXAMPLE.COM_
+----
+** Check that with the TGT, user authentication is successfully redirected:
```
```suggestion
** Verify user authentication by using your TGT:
```

maximiliankolb reviewed (changes requested):
some suggestions; the only necessary change is avoiding "foreman-maintain packages install" on Foreman Server.

asteflova commented:
I asked @mmuehlfeldRH to re-review.

mmuehlfeldRH commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,122 @@
+# mkdir /etc/ipa
+----
++
+.. Create a file named `default.conf` in the `/etc/ipa/` directory:
```
I would drop this step and, instead, change the next step to "Create the `/ettc/ipa/default.conf` file with the following content to configure the Kerberos realm for the AD domain:"

asteflova replied:
Addressed.

mmuehlfeldRH commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,122 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# mkdir /etc/ipa
```
```suggestion
# mkdir /etc/ipa/
```

to be aligned with the lead-in sentence of this code block.

mmuehlfeldRH commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,122 @@
+realm = _AD.EXAMPLE.COM_
+----
+. Configure the Apache keytab for Kerberos connections:
+.. Create a file named `/etc/net-keytab.conf` to store configuration details for Samba:
```
Same as above. A step to create an empty file seems superfluous if the next step is "edit the file". An editor can create new files. :-)

mmuehlfeldRH replied:
Is there a reason why not /etc/samba/smb.conf? 
Isn't there a risk that someone creates smb.conf later and then things get mixed up?

asteflova replied:
I merged the steps.

asteflova replied:
About smb.conf: That's very good insight. I'll test the procedure with smb.conf to see if it works that way too.

asteflova replied:
It works! The procedure now instructs users to store the configuration in the default /etc/samba/smb.conf and this file is then used to generate the keytab.

mmuehlfeldRH commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,122 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# KRB5_KTNAME=FILE:/etc/httpd/conf/http.keytab net ads keytab add HTTP -U Administrator -s /etc/net-keytab.conf -d3
```
Is debug level 3 (-d3) intentional?

asteflova replied:
It is. There is no output if creating the keytab fails and it took me a while to figure out what was happening. Level 3 debugging was used in the legacy procedure, perhaps a different level would suffice..? I'll try to tweak this when I re-test the procedure and update if necessary.

asteflova replied:
I checked the [man pages](https://www.samba.org/samba/docs/current/man-html/net.8.html) and anything higher than 1 is indeed supposed to be used for troubleshooting. So I dropped it after all.

mmuehlfeldRH commented on guides/common/modules/proc_configuring-the-active-directory-authentication-source-on-projectserver.adoc:
```diff
@@ -0,0 +1,122 @@
+----
+
+.Verification
+* To verify that AD users can log in to {ProjectWebUI}, log in to {ProjectwebUI} by entering the credentials of a user defined in AD.
```
The duplicate "log in to {ProjectwebUI} makes the sentence hard to read (at least for me).

mmuehlfeldRH replied:
Is it not necessary to obtain a Kerberos ticket first?

asteflova replied:
> The duplicate "log in to {ProjectwebUI} makes the sentence hard to read (at least for me).

Good point, I simplified this.

asteflova replied:
> Is it not necessary to obtain a Kerberos ticket first?

Not here. In the first verification step, users can verify that they can log in with user name and password -- no kinit needed.

In the second verification step, they verify Kerberos SSO.

asteflova commented:
@mmuehlfeldRH: I addressed your suggestions and resolved the related comment threads.

@maximiliankolb: I addressed the one breaking change you found (installing packages) and resolved the other comments.

@ekohl: I believe I addressed your request for changes as well by dropping the steps related to privilege separation for Apache and verifying that the remaining steps work.

asteflova commented:
@adamruzicka and/or @lhellebr: Is the latest state something you'd be comfortable approving?

adamruzicka commented:
I've tested this prior to the samba changes and back then it worked well. With the samba changes it reads well, so +1 from me, even though I didn't have a chance to test it out end to end.

asteflova commented:
Thanks for confirming @adamruzicka! With my testing, I believe this covers tech review.

asteflova commented:
There are still pending change requests but I checked them and they are either resolved or I have provided an explanation. Comment threads have been resolved. I'm going to go ahead with the merge now. Thanks everyone who provided their input!

asteflova commented:
Merged to "master" and cherry-picked:

1d6a8d3b1d..3d5e5d4295  3.12 -> 3.12
0d40788422..0ddfeaa68d  3.11 -> 3.11
